README.md
matlab_save_presentation
==============

Tools to create OpenOffice/LibreOffice and PowerPoint (PPTX) presentations from Matlab.


This is temporarily split into two directories. This is bound to change.


Features
--------

- Writes arbitrary figures/diagrams (like _saveas_)
- Append to existing presentation, or replace
- Does not require OpenOffice/LibreOffice to be installed
- Output conforms to the OASIS document standard
- Does not use COM or other IPC 

Note that ODP presentations can be converted to PPT with the [unoconv](http://dag.wiee.rs/home-made/unoconv/) software (but it requires OO/LO to be installed).


Usage
-----

```
saveodp(h,filename,['-a'])
```

Saves the figure in the file handle _h_ into the OO presentation _filename_. The file is appended to if '-a' is passed as an argument.


Installation
------------

The software is composed by two parts: 
- the ```saveodp.m``` file, which you should copy anywhere in Matlab's path, and 
- the ```img2odp.pl``` script, which should be in the executable path. 

The script relies on a number of Perl modules to work. Installation is straightforward under Linux, less so under Windows.  For this reason,  a standalone-executable version is provided.

*Linux.* Use your distribution's package manager or the *cpan* command to install the _OpenOffice::OODoc_ module.
 

Troubleshooting
---------------

If you get...

```
Error using saveodp (line 13)
Can't locate OpenOffice/OODoc.pm in @INC (@INC contains: /home/toni/perl5/lib/perl5/i386-linux-thread-multi
/home/toni/perl5/lib/perl5/i386-linux-thread-multi /home/toni/perl5/lib/perl5 /usr/local/lib/perl5 /usr/local/share/perl5
/usr/lib/perl5/vendor_perl /usr/share/perl5/vendor_perl /usr/lib/perl5 /usr/share/perl5 .) at img2odp.pl line 13.
BEGIN failed--compilation aborted at img2odp.pl line 13.
```

...you need to install the OpenOffice::OODoc Perl module. Try e.g. to run ```cpan``` then ```install OpenOffice::OODoc```. Answer _yes_ to all questions.


License
-------

    matlab_saveodp - Create OpenOffice/LibreOffice presentations from Matlab.
    Copyright (C) 2014 

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.


ChangeLog
2014-05-27  Toni  <toni@tabularasa>

	* 0.2 Workaround the "need invoke twice" bug

2014-05-25  Toni  <toni@tabularasa>

	* 0.1 Initial release



odftoolkit-0.6-incubating/Test.java
/*
  cp=`echo *.jar | sed 's/ /:/g'`

  javac -cp odfdom-java-0.8.9-incubating.jar:simple-odf-0.8-incubating.jar:xslt-runner-1.2.2-incubating.jar:xslt-runner-task-1.2.2-incubating.jar Test.java 

  java -cp $cp:. Test

  java -cp $cp -jar /usr/share/java/js.jar
*/



import java.io.File;
import java.net.*;

import org.odftoolkit.simple.*;
import org.odftoolkit.simple.presentation.*;
import org.odftoolkit.simple.presentation.Slide.*;

/**
 * This class is a demo to show the presentation API of Simple Java API for ODF.
 * 
 *
 */
public class Test {

	public static void main(String[] args) throws Exception {
	    PresentationDocument document = PresentationDocument.newPresentationDocument();
	    int n=document.getSlideCount();
	    document.newSlide(n, "new slide", null);
	    URI imageuri=new URI("../x.png");
	    document.newImage(imageuri);
	    document.save("shit.odp");
	}
	
}



pptx/savepptx.m

function savepptx(h,pptxfile,varargin)
% SAVEPPTX Create Office Open XML presentations from Matlab.
%
%   savepptx(h,pptxfile,['-a']) saves figure handle h in a PPTX file.
%
%   Appends figure as a separate slide at the end of the presentation
%   if the -a option is present.
narginchk(2,3);

setjavapath; 

% Eval nonsense is an attempt to workaround this bug http://www.mathworks.com/matlabcentral/answers/48233-how-come-javaaddpath-import-do-not-work-in-a-script-function
eval('import org.apache.poi.xslf.usermodel.*');

if(nargin==3)
    ppt=XMLSlideShow(java.io.FileInputStream(pptxfile));
else
    ppt=XMLSlideShow();
    ppt.setPageSize(java.awt.Dimension(1200,900));
end

slide=ppt.createSlide();
tn=[tempname '.png'];
saveas(h,tn,'png');

pictureData = org.apache.commons.io.IOUtils.toByteArray(java.io.FileInputStream(tn));
idx = ppt.addPicture(pictureData, XSLFPictureData.PICTURE_TYPE_PNG);
pic = slide.createPicture(idx);

out = java.io.FileOutputStream(pptxfile);
ppt.write(out);
out.close();

delete(tn);

end


function setjavapath()

% for some reason global and persistent do not work
% Run only once
if(any(cell2mat(regexp(javaclasspath,'poi-ooxml'))))
    return
end

[here,~,~] = fileparts(mfilename('fullpath'));
poidir=fullfile(here,'poi-3.10-FINAL/');
for i={'ooxml-lib/xmlbeans-2.3.0.jar'   ...
        'ooxml-lib/dom4j-1.6.1.jar'   ...
        'poi-3.10-FINAL-20140208.jar'   ...
        'poi-ooxml-3.10-FINAL-20140208.jar'  ...
        'poi-ooxml-schemas-3.10-FINAL-20140208.jar'}
    jar=fullfile(poidir,i{:});
    javaaddpath(jar);
end

end


img2ooImpress.pl
#!/usr/bin/perl -w

# T Giorgino May 2014 

# creates a Open Office Impress presentation from
# a list of images (assuming full page images)

# Inspired by a script by K.-H. Herrmann June 2005




use OpenOffice::OODoc;
use Getopt::Std;

our $opt_o="img2ooImpressExport.odp";
our $opt_a='';

getopts('o:a') or die "Parsing command line\n";

if (@ARGV == 0) {
  print << "EOF";
  Usage:
  img2ooImpress.pl [-a] [-o outputfile.odp] Image1.jpg Image2.jpg ...
     -a   Append to existing presentation (do not overwrite)
     -o   Output file
EOF

exit;
}

my $cont;
if($opt_a && -r $opt_o) {
	print "Appending\n";
	$cont=odfDocument(file=>$opt_o,part=>'content');
} else {
	print "Creating\n";
	$cont=odfDocument(file=>$opt_o,part=>'content',create=>'presentation');
}

$cont or die "Couldn't create document: $!\n";

my $i=1;
while (my $imgfile=shift @ARGV){
    my $test=$cont->insertDrawPage(-1);
    my $image=  $cont->createImageElement
	(
	 "Slide".$i." ".localtime,
	 description     => "image ".$i." filename:".$imgfile,
	 page            => $test,
	 position        => "0,0",
	 import          => $imgfile,
	 size            => "28cm, 21cm",
	 style           => "slide"
	);
    $i++;
}

$cont->save;



odp/saveodp.m

function saveodp(h,pptxfile,varargin) 
% SAVEODP Create Office Open XML presentations from Matlab. 
%
%   saveodp(h,pptxfile,['-a']) saves figure handle h in a PPTX file.   
%
%   Appends figure as a separate slide at the end of the presentation 
%   if the -a option is present. 
    narginchk(2,3);

    setjavapath;
    import org.odftoolkit.simple.*;
 
    if(nargin==3)
        ppt=PresentationDocument.loadDocument(pptxfile)
    else
        ppt=PresentationDocument.newPresentationDocument();  
      %  ppt.setPageSize(java.awt.Dimension(1200,900));
    end
    
    slide=ppt.newSlide(-1,'No name',SlideLayout.BLANK);
    tn=[tempname '.png'];
    saveas(h,tn,'png');
    
    pictureData = org.apache.commons.io.IOUtils.toByteArray(java.io.FileInputStream(tn));
    idx = ppt.addPicture(pictureData, XSLFPictureData.PICTURE_TYPE_PNG);
    pic = slide.createPicture(idx);

    out = java.io.FileOutputStream(pptxfile);
    ppt.write(out);
    
    delete(tn);

end
 


function setjavapath()
    [here,~,~] = fileparts(mfilename('fullpath'));
    poidir=fullfile(here,'odftoolkit-0.6-incubating/');
    javaaddpath(fullfile(poidir, 'odfdom-java-0.8.9-incubating.jar'));
    javaaddpath(fullfile(poidir, 'simple-odf-0.8-incubating.jar'));
    javaaddpath(fullfile(poidir, 'xslt-runner-1.2.2-incubating.jar'));
    javaaddpath(fullfile(poidir, 'xslt-runner-task-1.2.2-incubating'));
end
